packages/discord.js/src/client/Client.js

```diff
@@ -233,13 +275,149 @@ class Client extends BaseClient {
+    // * This can be optionally changed in the client options via the `waitGuildTimeout` option
+    // * a timeout time of zero will skip this timeout, which potentially could cause the Client to miss guilds.
+
+    this.readyTimeout = setTimeout(
+      () => {
+        this.emit(
+          Events.Debug,
+          `${
+            hasGuildsIntent
+              ? `Client did not receive any guild packets in ${this.options.waitGuildTimeout} ms.`
+              : 'Client will not receive anymore guild packets.'
+          }\nUnavailable guild count: ${this.expectedGuilds.size}`,
+        );
+
+        this.readyTimeout = null;
+        this.status = Status.Ready;
+
+        this._triggerClientReady();
+      },
+      hasGuildsIntent ? this.options.waitGuildTimeout : 0,
+    ).unref();
+  }
+
+  /**
+   * Attaches event handlers to the WebSocketShardManager from `@discordjs/ws`.
+   * @private
+   */
+  _attachEvents() {
+    this.ws.on(WebSocketShardEvents.Debug, (message, shardId) =>
+      this.emit(Events.Debug, `[WS => ${typeof shardId === 'number' ? `Shard ${shardId}` : 'Manager'}] ${message}`),
+    );
+    this.ws.on(WebSocketShardEvents.Dispatch, this._handlePacket.bind(this));
+
+    this.ws.on(WebSocketShardEvents.Ready, data => {
+      this.expectedGuilds = new Set([...this.expectedGuilds, ...data.guilds.map(guild => guild.id)]);
+      this.status = Status.WaitingForGuilds;
+      this._checkReady();
+    });
+
+    this.ws.on(WebSocketShardEvents.HeartbeatComplete, ({ heartbeatAt, latency }, shardId) => {
+      this.emit(Events.Debug, `[Shard ${shardId}] Heartbeat acknowledged, latency of ${latency}ms.`);
+      this.lastPingTimestamp = heartbeatAt;
+      this.pings.set(shardId, latency);
+    });
+  }
+
+  /**
+   * Processes a packet and queues it if this WebSocketManager is not ready.
+   * @param {Object} [packet] The packet to be handled
+   * @param {number} [shardId] The shardId that received this packet
+   * @private
+   */
+  _handlePacket(packet, shardId) {
+    if (packet && this.status !== Status.Ready && !BeforeReadyWhitelist.includes(packet.t)) {
+      this.packetQueue.push({ packet, shardId });
+    } else {
+      if (this.packetQueue.length) {
+        const item = this.packetQueue.shift();
+        setImmediate(() => {
+          this._handlePacket(item.packet, item.shardId);
+        }).unref();
+      }
+
+      if (packet && PacketHandlers[packet.t]) {
+        PacketHandlers[packet.t](this, packet, shardId);
+      }
+
+      if (this.status === Status.WaitingForGuilds && WaitingForGuildEvents.includes(packet.t)) {
+        this.expectedGuilds.delete(packet.d.id);
+        this._checkReady();
+      }
+    }
+  }
+
+  /**
+   * Broadcasts a packet to every shard of this client handles.
+   * @param {Object} packet The packet to send
+   * @private
+   */
+  _broadcast(packet) {
+    this.ws.getShardIds().then(shardIds => {
+      for (const shardId of shardIds) this.ws.send(shardId, packet);
```

Comment:
Is it intended this `_broadcast` method resolves before the work is done? `send` is "maybe async", but with current implementations it always is. Don't we want like a `Promise.all`-esque thing going on here?

Reply:
`_broadcast` is only really used in `ClientUser#setPresence()`, which immediately patches the presence locally. I'm fine with turning all of those into async to be awaitable along the chain, but that means if one shard fails to send the presence we end with... what exactly? Because some shards successfully set the presence while others didn't. Not to mention that our ClientPresence class doesn't support having different presences per shard stored locally.

Reply:
> Because some shards successfully set the presence while others didn't.

What do you mean by this? `send` never throws (unless the throttler/whatever has a bug). if setting the presence "fails" it's either silently dropped by Discord or we get an error event

Reply:
Then what good would awaiting the `send` do? As far as I can tell all that would do would be waiting for the websocket ratelimit to pass if we currently had hit the limit. Because we don't return/await the actual websocket sending in it anyway. Unless that is a sync operation which I hope it isn't.

Reply:
Yeah, exactly, it would wait for the message to have actually been sent. I feel like it's just bad design to have an async method that queues load you cannot possibly wait for. We should make it so the user can *choose* to wait (for whatever reason, like they *want* to make sure the presence updates were sent during startup) and as such, `await` whatever calls `_broadcast` (in this case setPresence) or.. not, if they don't care. cc @vladfrangu if you have diff. thoughts

Reply:
The thing is that the `ClientUser#setPresence()` call returns the ClientPresence which gets immediately patched locally anyway. Making that async would mean we'd have a `Promise<ClientPresence>` although the return value doesn't change after the initial sync call...
And as I said awaiting the `WebSocketManager#send()` call wouldn't make sure the payload actually was sent because it's not awaiting (and can't await) the `WebSocket#send()` call (because the WebSocket API doesn't support it). So if anything making them async/await would give the users a false sense of being able to actually wait for the payload to be sent successfully.

Reply:
Re that last part, while true, it's about the queue API we have. Regardless though, I digress. You make a fair point as well, let's let others decide

Reply:
Splitting my answers into two parts:
- _broadcast, should it stay, should become async/await and properly handle that
- ClientUser#setPresence becoming async is 🤷‍♂️. Imo we can either make it so (in which case that'll cascade for setActivity, and technically make it consistent with other methods), or call `void broadcast`. Both are fine _with me_, but maybe others disagree. @discordjs/core ☝️ thoughts?

Reply:
Since _broadcast literally only is used for ClientPresence#set I'll just make all of them async/await. Users of discord.js can still decide whether they want to await it or not i guess but you convinced me that it's the better approach.